### PR TITLE
cluster endpoints conflicts when creating internal serviceentry

### PR DESCRIPTION
**Pull request by ChangyuWang**
1. set **default_service_export_to: [.]** in configmap istio

2. httpbin is a internal service in namespace bar with a endpoint: ip 172.17.0.20 and port 5000, to access httpbin from foo, I created a internal serviceentry in foo
    ```
    apiVersion: networking.istio.io/v1alpha3
    kind: ServiceEntry
    metadata:
      name: internal-svc-mesh
      namespace: bar
    spec:
      hosts:
      - httpbin.bar.svc.cluser.local
      location: MESH_INTERNAL
      ports:
      - number: 80
        name: http
        protocol: HTTP
      resolution: DNS
    ```

    I found pilot will generate a outbound cluster "outbound|80||httpbin.bar.svc.cluser.local" for sidecar in foo below:
    ```
    {
      "name": "outbound|80||httpbin.bar.svc.cluster.local",
      "type": "STRICT_DNS",
      "connect_timeout": "1s",
      "circuit_breakers": {
       "thresholds": [
        {
         "max_retries": 1024
        }
       ]
      },
      "dns_refresh_rate": "5s",
      "dns_lookup_family": "V4_ONLY",
      "load_assignment": {
       "cluster_name": "outbound|80||httpbin.bar.svc.cluster.local",
       "endpoints": [
        {
         "lb_endpoints": [
          {
           "endpoint": {
            "address": {
             "socket_address": {
              "address": "httpbin.bar.svc.cluster.local",
              "port_value": 80
             }
            }
           },
           "load_balancing_weight": 1
          },
          {
           "endpoint": {
            "address": {
             "socket_address": {
              "address": "172.17.0.20",
              "port_value":5000
             }
            }
           },
           "load_balancing_weight": 1
          }
         ]
        }
       ]
      }
     }
    ```

With cluster "outbound|80||httpbin.bar.svc.cluser.local" configuration, there are two **Questions** here:
1. Endpoint with address 172.17.0.20 and port 5000 is conflict with resolution dns
2. Envoy applys cds failed with such configuration in ipv6 scenario even if i changed dns_lookup_family to V6_ONLY

### CODE REVIEW

**Comment by howardjohn** [comment]
/ok-to-test

Thanks for finding this, looks like a good bug fix. Can you add some tests for this so we don't run into regressions in the future?

**Comment by howardjohn** [comment]
/hold

somehow picked up 40 commits in this change, I think a rebase went wrong

**Comment by ChangyuWang** [comment]
if have fixed it

**Comment by howardjohn** [comment]
I built this locally and still saw the issues: `"EDS ACK ERROR: malformed IP address: httpbin.locality.svc.cluster.local. Consider setting resolver_name or setting cluster type to 'STRICT_DNS' or 'LOGICAL_DNS'`

**Comment by ChangyuWang** [comment]
> I built this locally and still saw the issues: `"EDS ACK ERROR: malformed IP address: httpbin.locality.svc.cluster.local. Consider setting resolver_name or setting cluster type to 'STRICT_DNS' or 'LOGICAL_DNS'`

when build outbound cluster of type "EDS", only DNSLB allowed, then this cluster name pass throuth eds interface. In the step of pushEds, pilot adds all endpoints of the releated service to envoy endpoints. I think it is a bug, the endpoint address with FQDN should be removed.

File: pilot/pkg/model/validation.go [hunk 1]
```diff
@@ -2478,18 +2478,27 @@ func appendErrors(err error, errs ...error) error {
...
+}
+
+// validateAddress checks ip or unix address
+func validateAddress(family AddressFamily, address string) error {
+	switch family {
 	case AddressFamilyTCP:
-		ipAddr := net.ParseIP(n.Address) // Typically it is an IP address
+		ipAddr := net.ParseIP(address) // Typically it is an IP address
 		if ipAddr == nil {
-			if err := ValidateFQDN(n.Address); err != nil { // Otherwise could be an FQDN
```
Comment (rshriram) [inline pilot/pkg/model/validation.go]: why remove fqdn?

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 2]
```diff
@@ -357,6 +357,9 @@ func buildLocalityLbEndpoints(env *model.Environment, proxyNetworkView map[strin
 			// Endpoint's network doesn't match the set of networks that the proxy wants to see.
 			continue
 		}
+		if service.Resolution != instance.Service.Resolution {
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: how can this be different?

**Comment by rshriram** [comment]
@howardjohn falls under service entry in multiple namespaces.. and raises bunch of other issues that I talked about.

File: pilot/pkg/model/validation.go (hunk 1 above)
Comment (ChangyuWang) [reply pilot/pkg/model/validation.go]: As described in [envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/endpoint/endpoint.proto#envoy-api-msg-endpoint-endpoint), for EDS, the address of endpoint is expected to be a direct IP address. Actually, if I create a serviceentry below, whose host is  already existed in cluster 
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: internal-svc-mesh
  namespace: bar
spec:
  hosts:
  - httpbin.bar.svc.cluser.local
  location: MESH_INTERNAL
  ports:
  - number: 80
    name: http
    protocol: HTTP
  resolution: DNS
```
envoy will report a error: 
```
error_detail:<code:13 message:"malformed IP address: httpbin.bar.svc.cluster.local. Consider setting resolver_name or setting cluster type to 'STRICT_DNS' or 'LOGICAL_DNS'" >
```

File: pilot/pkg/networking/core/v1alpha3/cluster.go (hunk 2 above)
Comment (ChangyuWang) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: like top described

File: pilot/pkg/model/validation.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/model/validation.go]: Yeah but what if you don't have `httpbin.bar.svc.cluser.local` already in your cluster? Then this service entry is perfectly valid

Comment (ChangyuWang) [reply pilot/pkg/model/validation.go]: Under normal circumstances, it works well. But I think such case should be taken into consideration, or the whole mesh would crash.

Comment (howardjohn) [reply pilot/pkg/model/validation.go]: sure but this breaks a bunch as well. How can I define a service entry for an external service now ?

Comment (ChangyuWang) [reply pilot/pkg/model/validation.go]: I have run e2e locally, and there is no impact on external service

**Pull request closed without merging**